File: generator/src/test/java/quan/generator/message/role/SRoleLogin.java
package quan.generator.message.role;

import quan.message.Buffer;
import java.util.*;
import quan.generator.message.user.UserInfo;
import quan.message.Message;
import java.io.IOException;

/**
 * 角色登录
 * Created by 自动生成
 */
public class SRoleLogin extends Message {

    private long roleId = 0L;//角色id

    private String roleName = "";//角色名

    private RoleInfo roleInfo = new RoleInfo();//角色信息

    private ArrayList<RoleInfo> roleInfoList = new ArrayList<>();//角色信息

    private HashSet<RoleInfo> roleInfoSet = new HashSet<>();//角色信息

    private HashMap<Long, RoleInfo> roleInfoMap = new HashMap<>();//角色信息

    private UserInfo userInfo;//用户信息

    public SRoleLogin() {
        super(111);
    }

    public long getRoleId() {
        return roleId;
    }

    public void setRoleId(long roleId) {
        this.roleId = roleId;
    }

    public String getRoleName() {
        return roleName;
    }

    public void setRoleName(String roleName) {
        if (roleName == null){
            throw new NullPointerException();
        }
        this.roleName = roleName;
    }

    public RoleInfo getRoleInfo() {
        return roleInfo;
    }

    public void setRoleInfo(RoleInfo roleInfo) {
        if (roleInfo == null){
            throw new NullPointerException();
        }
        this.roleInfo = roleInfo;
    }

    public ArrayList<RoleInfo> getRoleInfoList() {
        return roleInfoList;
    }

    public HashSet<RoleInfo> getRoleInfoSet() {
        return roleInfoSet;
    }

    public HashMap<Long, RoleInfo> getRoleInfoMap() {
        return roleInfoMap;
    }

    public UserInfo getUserInfo() {
        return userInfo;
    }

    public void setUserInfo(UserInfo userInfo) {
        this.userInfo = userInfo;
    }

    @Override
    public SRoleLogin create() {
        return new SRoleLogin();
    }

    @Override
    public void encode(Buffer buffer) throws IOException {
        super.encode(buffer);

        buffer.writeLong(roleId);
        buffer.writeString(roleName);
        roleInfo.encode(buffer);

        buffer.writeInt(roleInfoList.size());
        for (RoleInfo roleInfoListValue : roleInfoList) {
            roleInfoListValue.encode(buffer);
        }

        buffer.writeInt(roleInfoSet.size());
        for (RoleInfo roleInfoSetValue : roleInfoSet) {
            roleInfoSetValue.encode(buffer);
        }

        buffer.writeInt(roleInfoMap.size());
        for (long roleInfoMapKey : roleInfoMap.keySet()) {
            buffer.writeLong(roleInfoMapKey);
            roleInfoMap.get(roleInfoMapKey).encode(buffer);
        }

        buffer.writeBool(userInfo != null);
        if (userInfo != null) {
            userInfo.encode(buffer);
        }

    }

    @Override
    public void decode(Buffer buffer) throws IOException {
        super.decode(buffer);

        roleId = buffer.readLong();
        roleName = buffer.readString();
        roleInfo.decode(buffer);

        int roleInfoListSize = buffer.readInt();
        for (int i = 0; i < roleInfoListSize; i++) {
            RoleInfo roleInfoListValue = new RoleInfo();
            roleInfoListValue.decode(buffer);
            roleInfoList.add(roleInfoListValue);
        }

        int roleInfoSetSize = buffer.readInt();
        for (int i = 0; i < roleInfoSetSize; i++) {
            RoleInfo roleInfoSetValue = new RoleInfo();
            roleInfoSetValue.decode(buffer);
            roleInfoSet.add(roleInfoSetValue);
        }

        int roleInfoMapSize = buffer.readInt();
        for (int i = 0; i < roleInfoMapSize; i++) {
            long roleInfoMapKey = buffer.readLong();
            RoleInfo roleInfoMapValue = new RoleInfo();
            roleInfoMapValue.decode(buffer);
            roleInfoMap.put(roleInfoMapKey, roleInfoMapValue);
        }

        if (buffer.readBool()) {
            if (userInfo == null) {
                userInfo = new UserInfo();
            }
            userInfo.decode(buffer);
        }

    }

    @Override
    public String toString() {
        return "SRoleLogin{" +
                "roleId=" + roleId +
                ",roleName='" + roleName + '\'' +
                ",roleInfo=" + roleInfo +
                ",roleInfoList=" + roleInfoList +
                ",roleInfoSet=" + roleInfoSet +
                ",roleInfoMap=" + roleInfoMap +
                ",userInfo=" + userInfo +
                '}';

    }

}

File: message/MessageCS/Test/Role/SRoleLogin.cs
using System;
using System.Collections.Generic;
using MessageCS;
using Buffer = MessageCS.Buffer;
using MessageCS.Test.User;

namespace MessageCS.Test.Role
{
	/// <summary>
	/// 角色登录<br/>
	/// Created by 自动生成
	/// </summary>
    public class SRoleLogin : Message
    {
        /// <summary>
		/// 角色id
		/// </summary>
		public long roleId { get; set; }

		private string _roleName = "";

        /// <summary>
		/// 角色名
		/// </summary>
		public string roleName
		{
	    	get => _roleName;
	    	set => _roleName = value ?? throw new NullReferenceException();
		}

		private RoleInfo _roleInfo = new RoleInfo();

        /// <summary>
		/// 角色信息
		/// </summary>
		public RoleInfo roleInfo
		{
	    	get => _roleInfo;
	    	set => _roleInfo = value ?? throw new NullReferenceException();
		}

        /// <summary>
		/// 角色信息
		/// </summary>
		public List<RoleInfo> roleInfoList { get; } = new List<RoleInfo>();

        /// <summary>
		/// 角色信息
		/// </summary>
		public HashSet<RoleInfo> roleInfoSet { get; } = new HashSet<RoleInfo>();

        /// <summary>
		/// 角色信息
		/// </summary>
		public Dictionary<long, RoleInfo> roleInfoMap { get; } = new Dictionary<long, RoleInfo>();

        /// <summary>
		/// 用户信息
		/// </summary>
		private UserInfo userInfo { get; set; }


		public SRoleLogin(): base(222)
		{
		}

		public override Message Create()
		{
			return new SRoleLogin();
		}

		public override void Encode(Buffer buffer)
		{
	    	base.Encode(buffer);

		    buffer.WriteLong(roleId);
		    buffer.WriteString(roleName);
		    roleInfo.Encode(buffer);

		    buffer.WriteInt(roleInfoList.Count);
		    foreach (var roleInfoList_Value in roleInfoList) {
				roleInfoList_Value.Encode(buffer);
		    }

		    buffer.WriteInt(roleInfoSet.Count);
		    foreach (var roleInfoSet_Value in roleInfoSet) {
				roleInfoSet_Value.Encode(buffer);
		    }

		    buffer.WriteInt(roleInfoMap.Count);
		    foreach (var roleInfoMap_Key in roleInfoMap.Keys) {
		        buffer.WriteLong(roleInfoMap_Key);
			    roleInfoMap[roleInfoMap_Key].Encode(buffer);
		    }

		    buffer.WriteBool(userInfo != null);
		    userInfo?.Encode(buffer);
		}

		public override void Decode(Buffer buffer)
		{
	    	base.Decode(buffer);

		    roleId = buffer.ReadLong();
		    roleName = buffer.ReadString();
		    roleInfo.Decode(buffer);

		    var roleInfoList_Size = buffer.ReadInt();
		    for (var i = 0; i < roleInfoList_Size; i++) {
			    var roleInfoList_Value = new RoleInfo();
			  	roleInfoList_Value.Decode(buffer);
			    roleInfoList.Add(roleInfoList_Value);
		    }

		    var roleInfoSet_Size = buffer.ReadInt();
		    for (var i = 0; i < roleInfoSet_Size; i++) {
			    var roleInfoSet_Value = new RoleInfo();
			  	roleInfoSet_Value.Decode(buffer);
			    roleInfoSet.Add(roleInfoSet_Value);
		    }

		    var roleInfoMap_Size = buffer.ReadInt();
		    for (var i = 0; i < roleInfoMap_Size; i++) {
			    var roleInfoMap_Key = buffer.ReadLong();
			    var roleInfoMap_Value = new RoleInfo();
				roleInfoMap_Value.Decode(buffer);
			    roleInfoMap.Add(roleInfoMap_Key, roleInfoMap_Value);
		    }

		    if (buffer.ReadBool()) {
		        if (userInfo == null) {
		            userInfo = new UserInfo();
		        }
		        userInfo.Decode(buffer);
            }
		}

		public override string ToString()
		{
			return "SRoleLogin{" +
					"roleId=" + roleId +
					",roleName='" + roleName + '\'' +
					",roleInfo=" + roleInfo +
					",roleInfoList=" + ToString(roleInfoList) +
					",roleInfoSet=" + ToString(roleInfoSet) +
					",roleInfoMap=" + ToString(roleInfoMap) +
					",userInfo=" + userInfo +
					'}';
		}
    }
}